combustion-android-ble/src/main/java/inc/combustion/framework/service/FoodSafeData.kt

```diff
@@ -0,0 +1,446 @@
+            val data = UByteArray(SIZE_BYTES)
+
+            val rawMode = when (this) {
+                is Simplified -> Mode.Simplified
+                is Integrated -> Mode.Integrated
+            }.toRaw
+
+            val rawProduct = when (this) {
+                is Simplified -> this.product.toRaw
+                is Integrated -> this.product.toRaw
+            }
+
+            data[0] =
+                rawMode or
+                (rawProduct.toUShort() shl 3).toUByte()
+
+            data[1] =
+                (rawProduct.toUShort() shr 5).toUByte() or
+                (serving.toRaw.toUShort() shl 5).toUByte()
+
+            val toPacked: (Double) -> UInt = { value -> (value / 0.05).toUInt() and 0x1FFFu }
+
+            val rawSelectedThreshold = toPacked(when (this) {
+                is Simplified -> 0.0
+                is Integrated -> this.completionCriteria.selectedThresholdReferenceTemperature
+            })
+            val rawZValue = toPacked(when (this) {
+                is Simplified -> 0.0
+                is Integrated -> this.completionCriteria.zValue
+            })
+            val rawReferenceTemperature = toPacked(when (this) {
+                is Simplified -> 0.0
+                is Integrated -> this.completionCriteria.referenceTemperature
+            })
+            val rawDValueAtRt = toPacked(when (this) {
+                is Simplified -> 0.0
+                is Integrated -> this.completionCriteria.dValueAtRt
+            })
+            val rawTargetLogReduction = when (this) {
+                is Simplified -> 0u
+                is Integrated -> (this.completionCriteria.targetLogReduction / 0.1).toUInt() and 0xFFu
+            }
+
+            data[2] =
+                (rawSelectedThreshold and 0b0_0000_1111_1111u).toUByte()
+
+            data[3] =
+                ((rawSelectedThreshold and 0b1_1111_0000_0000u) shr 8).toUByte() or
+                ((rawZValue and 0b0_0000_0000_0111u) shl 5).toUByte()
+
+            data[4] =
+                ((rawZValue and 0b0_0111_1111_1000u) shr 3).toUByte()
+
+            data[5] =
+                ((rawZValue and 0b1_1000_0000_0000u) shr 11).toUByte() or
+                ((rawReferenceTemperature and 0b0_0000_0011_1111u) shl 2).toUByte()
+
+            data[6] =
+                ((rawReferenceTemperature and 0b1_1111_1100_0000u) shr 6).toUByte() or
+                ((rawDValueAtRt and 0b0_0000_0000_0001u) shl 7).toUByte()
+
+            data[7] =
+                ((rawDValueAtRt and 0b0_0001_1111_1110u) shr 1).toUByte()
+
+            data[8] =
+                ((rawDValueAtRt and 0b1_1110_0000_0000u) shr 9).toUByte() or
+                ((rawTargetLogReduction and 0b0000_1111u) shl 4).toUByte()
+
+            data[9] =
+                (rawTargetLogReduction and 0b1111_0000u shr 4).toUByte()
+
+            return data
+        }
+
+    companion object {
+        const val SIZE_BYTES = 10
+
+        fun fromRawData(data: UByteArray): FoodSafeData? {
```

Comment:
Just an opinion -- it would be nice to keep this kind of code internaly scoped and alongside the rest of the BLE serialization/deserialization.  Maybe there is a way to use extension functions on `FoodSafeData` and `FoodSafeStatus` to achieve that encapsulation.

Reply:
Nevermind -- it looks like there are a bunch of serialization converters have public scope.  Probably no need to change.

Reply:
This is a good call--Matt and I discussed this and it seems like the happy medium between following existing convention and getting proper encapsulation is to mark all the de-/serialization methods `internal`.